### ISSUE TITLE
How long does it take to train on librispeech dataset?

### ISSUE THREAD

**Issue by ArtemiyFirsov** [SeanNaren/deepspeech.pytorch#114]
I finally configured everything and strted learning on librispeech dataset with parameters from https://arxiv.org/pdf/1512.02595.pdf

Now, I am curious, how long will it take to learn this dataset? Maybe you know how to estimate this time?

**Comment by mcernak** [comment]
I trained LibriSpeech 100h clean data model (with default parameters) and it took on one GPU about 20 hours (21 epochs, a single epoch in about 1 hour).

**Comment by ArtemiyFirsov** [comment]
@mcernak thank you. 
I watched Baidu DeepSpeech2 video https://www.youtube.com/watch?v=g-sndkf7mCs
There they explain how long does it take to learn and provide the formula:
![image](https://user-images.githubusercontent.com/4475110/28352330-fb5e841a-6c7f-11e7-87b4-f1cd9ae2661c.png)

connections = number of unique connections in RNN
frames = number of frames in unrolled RNN
utterances = number of utterances in dataset
epochs = number of epochs
3 = backprop + forward prop + gradient update
2 = for multiplying every connection

I guess that connections  is the number of parameters estimate. In this project, you may find that the number of parameters is printed to stdout, by default it was 11376368 (approx. 10^7)
Number of utterances in train libri = 562482 (approx. 10^6)
epochs = 70 (by default)
frames = do not yet understand this (how much frames is there in an unrolled RNN???). I guess it is RNN layer size, which by default 400. 

so, 10^7 * 4 * 10^2 * 10^6 * 70 * 3 * 2 = approx. 10^18, i.e. on 10 teraflops it should be 10^6 secs ideally ~ 270 hours for full dataset. 
in your case, you had 3.5 times less epochs and 10 times less data (approx), so you should had about 10 hours of training. in that prospective, 20 hours of training looks quite explainable))

**Comment by SiddGururani** [comment]
Good job finding this formula! There's a couple of things that I think you might be wrong about:
1. So the number printed to stdout isn't the number of RNN connections, it's the CNN+RNN+fully-connected weights all combined. I believe the 2-layer CNN with batch-norm contributes to a lot of parameters so you have to discount those.
2. The number of frames will vary from batch to batch and will be the max number of audio frames for the particular batch. You could probably take the mean/median to get a rough estimate.

**Comment by ArtemiyFirsov** [comment]
@SiddGururani 

1. Yes, you are right, but still those parameters are involved in learning, thus, adding some additional time in training
2. Yes, I guess, this is right assumption. However, mean should be quite OK for estimations.

**Comment by SiddGururani** [comment]
Oh you're right. I got a little nitpicky and assumed that 

> connections = number of unique connections in RNN

 meant only the RNN part of the network but it looks like you meant the complete model since all weights are going to be involved in the floating point operations. My bad haha!
